File: sections/14_core_patterns_mixed_contest/problems/c_active_colors/solution.py
"""Reference solution for C. Active Colors."""

import sys


def solve_case(intervals: list[tuple[int, int, int]], queries: list[int]) -> list[int]:
    events: list[tuple[int, int, int, int]] = []
    for l, r, color in intervals:
        events.append((r, 0, color, -1))
        events.append((l, 1, color, -1))
    for idx, x in enumerate(queries):
        events.append((x, 2, 0, idx))
    events.sort()

    freq: dict[int, int] = {}
    distinct = 0
    ans = [0] * len(queries)
    for _, kind, color, idx in events:
        if kind == 0:
            count = freq[color] - 1
            if count == 0:
                distinct -= 1
                del freq[color]
            else:
                freq[color] = count
        elif kind == 1:
            count = freq.get(color, 0)
            if count == 0:
                distinct += 1
            freq[color] = count + 1
        else:
            ans[idx] = distinct
    return ans


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        q = data[idx + 1]
        idx += 2
        intervals: list[tuple[int, int, int]] = []
        for _ in range(n):
            l = data[idx]
            r = data[idx + 1]
            color = data[idx + 2]
            idx += 3
            intervals.append((l, r, color))
        queries = data[idx:idx + q]
        idx += q
        out.append(" ".join(map(str, solve_case(intervals, queries))))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/c_active_colors/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Event {
    long long x;
    int kind;
    long long color;
    int idx;
};

vector<int> solve_case(const vector<tuple<long long, long long, long long>> &intervals, const vector<long long> &queries) {
    vector<Event> events;
    events.reserve(intervals.size() * 2 + queries.size());
    for (auto [l, r, color] : intervals) {
        events.push_back({r, 0, color, -1});
        events.push_back({l, 1, color, -1});
    }
    for (int i = 0; i < (int)queries.size(); ++i) {
        events.push_back({queries[i], 2, 0, i});
    }
    sort(events.begin(), events.end(), [](const Event &a, const Event &b) {
        if (a.x != b.x) return a.x < b.x;
        return a.kind < b.kind;
    });

    unordered_map<long long, int> freq;
    int distinct = 0;
    vector<int> ans(queries.size());
    for (const Event &event : events) {
        if (event.kind == 0) {
            int count = freq[event.color] - 1;
            if (count == 0) {
                --distinct;
                freq.erase(event.color);
            } else {
                freq[event.color] = count;
            }
        } else if (event.kind == 1) {
            int count = freq[event.color];
            if (count == 0) ++distinct;
            freq[event.color] = count + 1;
        } else {
            ans[event.idx] = distinct;
        }
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n, q;
        cin >> n >> q;
        vector<tuple<long long, long long, long long>> intervals;
        intervals.reserve(n);
        for (int i = 0; i < n; ++i) {
            long long l, r, color;
            cin >> l >> r >> color;
            intervals.push_back({l, r, color});
        }
        vector<long long> queries(q);
        for (long long &x : queries) cin >> x;
        vector<int> ans = solve_case(intervals, queries);
        for (int i = 0; i < q; ++i) {
            if (i) cout << ' ';
            cout << ans[i];
        }
        cout << '\n';
    }

    return 0;
}

